File: StringResourcesConverter/src/com/skedgo/tools/platform/xliff/XLIFFInputStrategy.java
package com.skedgo.tools.platform.xliff;

import java.io.InputStream;

import javax.xml.parsers.SAXParser;
import javax.xml.parsers.SAXParserFactory;

import org.xml.sax.Attributes;
import org.xml.sax.InputSource;
import org.xml.sax.SAXException;
import org.xml.sax.XMLReader;
import org.xml.sax.helpers.DefaultHandler;

import com.skedgo.tools.InputCreatorListener;
import com.skedgo.tools.InputStringsStrategy;
import com.skedgo.tools.model.StringDefinition;
import com.skedgo.tools.model.StringsStructure;
import com.sun.istack.internal.NotNull;

public class XLIFFInputStrategy implements InputStringsStrategy {

	private static XLIFFInputStrategy instance;

	private XLIFFInputStrategy() {
	}

	public static XLIFFInputStrategy getInstance() {
		if (instance == null) {
			instance = new XLIFFInputStrategy();
		}

		return instance;
	}

	
	public void createInputValues(InputStream input, InputCreatorListener listener, String fileSourceType) throws Exception {
		StringsStructure output = new StringsStructure();

		SAXParserFactory spf = SAXParserFactory.newInstance();
		spf.setNamespaceAware(true);
		SAXParser saxParser = spf.newSAXParser();
		XMLReader xmlReader = saxParser.getXMLReader();
		xmlReader.setContentHandler(new SAXLocalNameCount(output, listener, fileSourceType));
		xmlReader.parse(new InputSource(input));

	}
	
	@Override
	public void createInputValues(InputStream input, InputCreatorListener listener) throws Exception {
		createInputValues(input, listener, null);
	}
	

	class SAXLocalNameCount extends DefaultHandler {

		private StringsStructure structure;
		private InputCreatorListener listener;

		private boolean bSource;
		private boolean bNote;
		private boolean bTarget;
		private boolean bAltTrans;
		
		private boolean skipSource;

		private String source;
		private String note;
		private String target;
		
		private String file;

		public SAXLocalNameCount(@NotNull StringsStructure structure, @NotNull InputCreatorListener listener, String fileSourceType) {
			this.structure = structure;
			this.structure.setBaseStructure(new StringsStructure());
			this.listener = listener;
			this.file = fileSourceType;
		}

		public void startDocument() throws SAXException {

		}

		public void endDocument() throws SAXException {
			listener.didFinishInputCreation(structure);
		}

		public void startElement(String namespaceURI, String localName, String qName, Attributes atts)
				throws SAXException {

			if (qName.equals("file")) {
				structure.setTargetLanguage(atts.getValue("target-language"));
				skipSource = !(file == null || atts.getValue("original").endsWith(file));
			}

			if (qName.equalsIgnoreCase("trans-unit")) {
				source = null;
				note = null;
				target = null;
			}

			if (qName.equalsIgnoreCase("source")) {
				bSource = true;
			}
			if (qName.equalsIgnoreCase("note")) {
				bNote = true;
			}
			if (qName.equalsIgnoreCase("target")) {
				bTarget = true;
			}
			
			if (qName.equalsIgnoreCase("alt-trans")) {
				bAltTrans = true;
			}

		}

		@Override
		public void endElement(String uri, String localName, String qName) throws SAXException {

			if (!skipSource && qName.equalsIgnoreCase("trans-unit")) {
				if (note != null) {
					structure.getComments().put(structure.getDefinitions().size(), note);
					structure.getBaseStructure().getComments().put(structure.getDefinitions().size(), note);
				}
				if(target == null  || target.isEmpty()){
					structure.getDefinitions().put(structure.getDefinitions().size(), new StringDefinition(source, source));
					
				}else{
					structure.getDefinitions().put(structure.getDefinitions().size(), new StringDefinition(source, target));
					structure.getBaseStructure().getDefinitions().put(structure.getDefinitions().size(), new StringDefinition(source, source));
				}
			}
			

			bSource = false;
			bNote = false;
			bTarget = false;	
			bAltTrans = false;
			
			
		}

		public void characters(char ch[], int start, int length) throws SAXException {

			String characters = new String(ch, start, length);
			
			if (bSource) {
				if(source == null) source = "";
				source += characters;
				
			}
			if (bNote) {
				if(note == null) note = "";
				note += characters;
				
			}
			if (bTarget) {
				
				if(!bAltTrans)	{
					if(target == null) target = "";
					else target += " ";
					
					target += characters.trim();	
				}				
			}

		}

	}

}

File: src/main/java/com/skedgo/tools/platform/xliff/XLIFFInputStrategy.kt
package com.skedgo.tools.platform.xliff

import com.skedgo.tools.InputStringsStrategy
import com.skedgo.tools.model.TransUnit
import com.skedgo.tools.model.Translations
import org.xml.sax.Attributes
import org.xml.sax.InputSource
import org.xml.sax.helpers.DefaultHandler
import rx.Single
import rx.subjects.BehaviorSubject
import java.io.InputStream
import javax.xml.parsers.SAXParser
import javax.xml.parsers.SAXParserFactory

class XLIFFInputStrategy(var fileSourceType: String? = null) : InputStringsStrategy {

    val completed: BehaviorSubject<Translations> = BehaviorSubject.create()

    override fun createInputValues(input: InputStream): Single<Translations> =
            Single.fromCallable {
                val saxParser = initParser()
                parseInput(input, saxParser)
            }.flatMap { completed.first().toSingle() }

    private fun initParser(): SAXParser {
        val spf = SAXParserFactory.newInstance()
        spf.isNamespaceAware = true
        return spf.newSAXParser()
    }

    private fun parseInput(input: InputStream,
                           saxParser: SAXParser) {
        val xmlReader = saxParser.xmlReader
        xmlReader.contentHandler = ReaderHandler()
        xmlReader.parse(InputSource(input))
    }

    inner class ReaderHandler : DefaultHandler() {

        private val structure = Translations(mutableListOf())

        private var readingSource: Boolean = false
        private var readingNote: Boolean = false
        private var readingTarget: Boolean = false
        private var readingAltTrans: Boolean = false

        private var skipFile: Boolean = false

        private var id: String = ""
        private var source: String = ""
        private var note: String = ""
        private var target: String = ""

        override fun startDocument() {}

        override fun endDocument() {
            completed.onNext(structure)
        }

        override fun startElement(namespaceURI: String?, localName: String?, qName: String, atts: Attributes?) {
            when {
                qName == "file" && atts != null -> {
                    structure.targetLanguage = atts.getValue("target-language")
                    skipFile = shouldSkipFile(atts)
                }
                qName == "trans-unit" -> {
                    id = atts?.let { atts.getValue("id") } ?: ""
                    source = ""
                    note = ""
                    target = ""
                }
                qName == "source" -> readingSource = true
                qName == "note" -> readingNote = true
                qName == "target" -> readingTarget = true
                qName == "alt-trans" -> readingAltTrans = true
            }
        }

        private fun shouldSkipFile(atts: Attributes) =
                !(fileSourceType?.let { atts.getValue("original").endsWith(it) }
                        ?: true)

        override fun endElement(uri: String?, localName: String?, qName: String) {
            if (!skipFile && qName == "trans-unit") {
                structure.transUnits.add(TransUnit(id, source, resolveTarget(), note))
            }

            readingSource = false
            readingNote = false
            readingTarget = false
            readingAltTrans = false
        }

        private fun resolveTarget() =
                when (target.isEmpty()) {
                    true -> source
                    false -> target
                }

        override fun characters(ch: CharArray, start: Int, length: Int) {
            val characters = String(ch, start, length)
            when {
                readingSource -> source += characters
                readingNote -> note += characters
                readingTarget && !readingAltTrans -> target += characters.trim { it <= ' ' }
            }
        }
    }
}
